### PR TITLE
GA Review for ockam_transport_tcp crate

### PR DESCRIPTION
**Pull request by antoinevg**
### Checklist

* [x] 🤔 Should this file exist?
* [x] ❤️ Is the Public API friendly? 
* [x] 📚 Are all Public functions documented? 
* [x] 🪂 Do public functions have rust doc examples? 
* [x] 🆗 Do all functions return the correct errors?
* [x] 🛃 Are all error results returned inside the implementation being handled correctly?
* [x] ✅ Do all modules have sufficient test coverage?
* [x] 🪵 Is there sufficient logging and are the correct levels being used?
* [x] 🛌 `N/A` Are all modules ready for use on embedded?  
* [x] 📦 Are we using the correct Cargo features in all files and `Cargo.toml`?
* [x] ☣️ `N/A` Has all `unsafe` code been audited? 

### Open Questions

* 🤔 Should the buffer size in [`src/portal/portal_receiver.rs#L8`](https://github.com/ockam-network/ockam/blob/fe8e307f15474e10088f15da7831b168b2de299e/implementations/rust/ockam/ockam_transport_tcp/src/portal/portal_receiver.rs#L8) be hardcoded?
* 🤔 There are some outstanding TODO's in [`src/portal/portal_worker.rs#L148-L190`](https://github.com/ockam-network/ockam/blob/fe8e307f15474e10088f15da7831b168b2de299e/implementations/rust/ockam/ockam_transport_tcp/src/portal/portal_worker.rs#L148-L190)
* 🤔 Are processors affected by @spacekookie's recent work on `Processor` 

### New Tests

* ♻️ `src/lib.rs` - `test_parse_socket_address()`

### Refactored

* ♻️ `src/router/mod.rs` - Moved TcpRouter into its own file `tcp_router.rs`
* ❤️ `src/router/mod.rs` - friendlify code organisation of `ockam_transport_tcp::TcpRouter` 
* ❤️ `src/router/handle.rs` - friendlify code organisation of `ockam_transport_tcp::TcpRouterHandle`
* ❤️ `src/workers/listener.rs` - friendlify code organisation of `ockam_transport_tcp::TcpListenProcessor`
* ❤️ `src/workers/sender.rs` - friendlify code organisation of `ockam_transport_tcp::TcpSendWorker`

### CODE REVIEW

**Comment by antoinevg** [comment]
Grrf

File: implementations/rust/ockam/ockam_transport_tcp/src/workers/sender.rs
Comment (thomcc) [inline implementations/rust/ockam/ockam_transport_tcp/src/workers/sender.rs:217]: This is honestly just cuz of our encoding trait being bad. I indend to rework this eventually, and started on this at one point pre-GA

```rs
// Also, this should be renamed to `Encode` for better rust style...
pub trait Encodable {
    fn encode_into(&self, buf: &mut Encoded) -> Result<()>;
    fn encode(&self) -> Result<Encoded> {
        let mut buf = vec![];
        self.encode_into(&mut buf)?;
        Ok(buf)
    }
}

// And the blanket impl would be something like this
impl<T: Serialize> Encodable for T {
    fn encode_into(&self, buf: &mut Encoded) -> Result<()> {
        use serde_bare::ser::{Serializer, VecWrite};
        let mut serializer = Serializer::new(VecWrite::new(buf));
        self.serialize(&mut serializer)?;
        Ok(())
    }
}
```

But I haven't had time, and we'll be changing encoding formats eventually anyway. It also makes less sense with the `Encoded` type alias (which IMO probably should be removed anyway).

**Review by thomcc — APPROVED** [review]
This is great!

**Comment by thomcc** [comment]
Uh, hm. I was going to merge this now, but the conflicts from the rebase look... *really* hairy, so do you mind rebasing?

**Review by SanjoDeundiak — APPROVED** [review]

**Review by spacekookie — DISMISSED** [review]

**Comment by antoinevg** [comment]
> Uh, hm. I was going to merge this now, but the conflicts from the rebase look... _really_ hairy, so do you mind rebasing?

Hairy rebase was indeed hairy :-)

I've pushed it but we're now getting build errors during for the `Rust / Rust - Nightly *` checks:

```
   Compiling os_pipe v0.9.2
error[E0658]: trait bounds other than `Sized` on const fn parameters are unstable
   --> /usr/rust/cargo/registry/src/github.com-1ecc6299db9ec823/lock_api-0.4.7/src/mutex.rs:150:6
    |
150 | impl<R: RawMutex, T> Mutex<R, T> {
    |      ^
...
154 |     pub const fn new(val: T) -> Mutex<R, T> {
    |     --------------------------------------- function declared as const here
    |
    = note: see issue #93706 <https://github.com/rust-lang/rust/issues/93706> for more information
    = help: add `#![feature(const_fn_trait_bound)]` to the crate attributes to enable
```

I can't reproduce it locally so it looks like the `lock_api` sub-dependency pushed some changes which assumes a newer `nightly` than the CI machine:

```
rustup default nightly
  shell: sh -e {0}
info: using existing install for 'nightly-x86_64-unknown-linux-gnu'
info: default toolchain set to 'nightly-x86_64-unknown-linux-gnu'

  nightly-x86_64-unknown-linux-gnu unchanged - rustc 1.61.0-nightly (68369a041 2022-02-22)
```

Shouldn't `rustup` be checking and updating that automatically though? 

Do you have any advice @metaclips?

**Review by thomcc — DISMISSED** [review]
LGTM. The error in CI should be fixed by a rebase (which I'll do, since it shouldn't be so hairy)

**Comment by thomcc** [comment]
> since it shouldn't be so hairy

Should have kept my mouth shut -- I definitely spoke too soon. Please rebase (some changes are trivial but some I'm guessing at, and it would be a bad way to write a bug)

We should land this ASAP after you fix the rebase, since it seems very fragile.

**Comment by antoinevg** [comment]
> > since it shouldn't be so hairy
> 
> Should have kept my mouth shut -- I definitely spoke too soon. Please rebase (some changes are trivial but some I'm guessing at, and it would be a bad way to write a bug)
> 
> We should land this ASAP after you fix the rebase, since it seems very fragile.

Aurgh, there goes another morning. 😫

This branch should probably have been merged immediately after I rebased it on Wednesday, before merging the transport type changes:

https://github.com/ockam-network/ockam/commit/55ffce6af0ae1453d6901d192682bf70fa9edfc1

…which undoes all the work I did on the rebase.

What I'll try do is create a new branch off develop, back out the transport type changes, merge my changes and then rebase the transport type changes on top of mine as they are pretty mechanical.

**Comment by antoinevg** [comment]
Acke, just chatted with @SanjoDeundiak and we're going to have to do this a 3rd time if we're not careful! 🤣

How about:

1. I hold off on rebasing this branch until @SanjoDeundiak's tcp reconnect branch lands.
2. I do the hairy rebase
3. We put a freeze on landing anything else that touches `ockam_transport_tcp` until we've landed this ga review branch.
4. We review & land this branch
5. Back to business as usual

**Comment by thomcc** [comment]
It would be nice to get this in, since I think it's the one of the last bigish chunks of GA work that has not landed.

**Comment by thomcc** [comment]
I won't land anything that touches that code until then.

**Comment by antoinevg** [comment]
🚀 …and all your rebase are belong to us!  *\*touch.wood\** we should be good to go.

**Review by thomcc — APPROVED** [review]

**Pull request merged**
